### PR TITLE
Start OmniXtend Initialization Protocol Documentation

### PR DESCRIPTION
**Pull request by nategraff-sifive**
Here's how I currently best understand the plan for the OmniXtend Initialization Protocol, represented as Graphviz dot graphs. From here, we can discuss whether these graphs accurately represent the protocol we all had in mind, and hopefully converge on a design we're happy moving forward with.

This also forms the first content of the Initialization Protocol documentation.

As @daveparry instructed, I've licensed all of this content under the Creative Commons Attribution 4.0 International (CC BY 4.0) License.

Please add more reviewers if I missed anyone.

### CODE REVIEW

**Comment by atishp04** [comment]
Few comments: 
1. What exactly do you mean by change local hart ids ? mhartid is a read only CSR. 
I think you mean change the local hartid in DTB. In that case, it doesn't have to happen in individual nodes. Individual nodes will parse the updated DTB from the co-coordinator node to figure out the hartid instead of reading mhartid. 

2. Co-coordinator node can be a generic x86 node or a true omnixtend node. In the former case, it really doesn't have TLoE address maps & TLoE links. Those are only applicable if coordinator node is a true OmniXtend node. That's true for Config PHY and enable MAC as well.

3. Currently, there is no text in the spec. We should start writing text around the diagrams and include the diagrams in the asciidoc.

**Comment by nategraff-sifive** [comment]
Thanks for your feedback @atishp04!

>     1. What exactly do you mean by change local hart ids ? mhartid is a read only CSR.
>        I think you mean change the local hartid in DTB. In that case, it doesn't have to happen in individual nodes. Individual nodes will parse the updated DTB from the co-coordinator node to figure out the hartid instead of reading mhartid.

Okay, I wasn't clear on that, I thought that the hardware changes included a writable mhartid CSR. In either case, clearly each node is going to need to have some writable place to store the hartids is parses from the updated DTB.

>     2. Co-coordinator node can be a generic x86 node or a true omnixtend node. In the former case, it really doesn't have TLoE address maps & TLoE links. Those are only applicable if coordinator node is a true OmniXtend node. That's true for Config PHY and enable MAC as well.

Maybe we need 2 coordinator flowcharts: one for when the coordinator is not an OmniXtend node, and one when it is, since it's not necessarily clear that the two roles compose trivially.

>     3. Currently, there is no text in the spec. We should start writing text around the diagrams and include the diagrams in the asciidoc.

Happy to keep expanding with text :) I'm fully intending to start embedding these in asciidoc, I'm just pushing what I have as I have it.

**Comment by atishp04** [comment]
> Thanks for your feedback @atishp04!
> 
> > ```
> > 1. What exactly do you mean by change local hart ids ? mhartid is a read only CSR.
> >    I think you mean change the local hartid in DTB. In that case, it doesn't have to happen in individual nodes. Individual nodes will parse the updated DTB from the co-coordinator node to figure out the hartid instead of reading mhartid.
> > ```
> 
> Okay, I wasn't clear on that, I thought that the hardware changes included a writable mhartid CSR. 
In either case, clearly each node is going to need to have some writable place to store the hartids is parses from the updated DTB.
> 
That's not possible (writable mhartid CSR) as per the RISC-V spec. Yes. We will probably store it in scratch space. You can update the diagram accordingly.

> > ```
> > 2. Co-coordinator node can be a generic x86 node or a true omnixtend node. In the former case, it really doesn't have TLoE address maps & TLoE links. Those are only applicable if coordinator node is a true OmniXtend node. That's true for Config PHY and enable MAC as well.
> > ```
> 
> Maybe we need 2 coordinator flowcharts: one for when the coordinator is not an OmniXtend node, and one when it is, since it's not necessarily clear that the two roles compose trivially.

Sounds good to me. or just two branches in the same flow chart is also okay.
> 
> > ```
> > 3. Currently, there is no text in the spec. We should start writing text around the diagrams and include the diagrams in the asciidoc.
> > ```
> 
> Happy to keep expanding with text :) I'm fully intending to start embedding these in asciidoc, I'm just pushing what I have as I have it.

Awesome. Just update the PR as once you have some. I will review asap.

File: InitializationProtocol/ManagementProtocol.adoc [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+transitioned to the next stage of the boot flow.
+
+=== Glossary
+
+OmniXtend Cluster::
+	A collection of machines, connected by Ethernet, which includes both OmniXtend Nodes, an OmniXtend Coordinator, and optionally network switches.
+OmniXtend Node::
+	One of several machines in an OmniXtend Cluster which implements the TileLink over Ethernet protocol (TLoE).
+OmniXtend Coordinator::
+	A single machine in the OmniXtend Cluster which performs the Coordinator role in the OmniXtend Management Protocol. The Coordinator may be any machine connected to the same Ethernet network capable of sending and receiving Ethernet packets so that the Coordinator role may be performed by any of a separate physical server, a programmable network switch, or an OmniXtend Node.
```
Comment (atishp04) [inline InitializationProtocol/ManagementProtocol.adoc:19]: Rename it to management server.

File: InitializationProtocol/ManagementProtocol.adoc [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	A coherent chip-scale interconnect designed for use in a System-on-Chip to connect general-purpose multiprocessors, co-processors, accelerators, DMA engines, and other devices.
+Devicetree::
+	A markup specification for describing the hardware resources of a computer.
+Devictree Binary::
+	A particular encoding of Devicetree markup designed for machine consumption.
+Single-Node DTB::
+	A Devicetree in Devicetree Binary format which describes the hardware resources of an individual OmniXtend Node
+Cluster DTB::
+	A Devicetree in Devicetree Binary format which describes the hardware resources of the OmniXtend Cluster and contains all of the information necessary for all OmniXtend Nodes to configure and bring-up their TLoE hardware interfaces.
+
```
Comment (atishp04) [inline InitializationProtocol/ManagementProtocol.adoc:34]: OmniXtend topology node also need to be added here

File: InitializationProtocol/ManagementProtocol.adoc [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+TLoE::
+	An Acronym for TileLink over Ethernet
+TileLink::
+	A coherent chip-scale interconnect designed for use in a System-on-Chip to connect general-purpose multiprocessors, co-processors, accelerators, DMA engines, and other devices.
+Devicetree::
+	A markup specification for describing the hardware resources of a computer.
+Devictree Binary::
+	A particular encoding of Devicetree markup designed for machine consumption.
+Single-Node DTB::
+	A Devicetree in Devicetree Binary format which describes the hardware resources of an individual OmniXtend Node
```
Comment (atishp04) [inline InitializationProtocol/ManagementProtocol.adoc:31]: Let's call it local DTB. It represents the DTB that board booted with.

File: InitializationProtocol/ManagementProtocol.adoc [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+Cluster DTB::
+	A Devicetree in Devicetree Binary format which describes the hardware resources of the OmniXtend Cluster and contains all of the information necessary for all OmniXtend Nodes to configure and bring-up their TLoE hardware interfaces.
+
+== Management Protocol
+
+This section describes the protocol by which OmniXtend Nodes communicate over Ethernet with an OmniXtend Coordinator to:
```
Comment (atishp04) [inline InitializationProtocol/ManagementProtocol.adoc:37]: The flow is different for booting node & non-booting node depending on whether NUMA system or distributed system is used. We need to clarify those upfront.We also need to add another diagram for booting node.

File: InitializationProtocol/ManagementProtocol.adoc (hunk 1 above)
Comment (nategraff-sifive) [reply InitializationProtocol/ManagementProtocol.adoc:19]: Two hard problems in computer science (cache invalidation, naming things, and off by one errors) and this project encompasses all of them. 😅

File: InitializationProtocol/ManagementProtocol.adoc (hunk 4 above)
Comment (nategraff-sifive) [reply InitializationProtocol/ManagementProtocol.adoc:37]: Once the TLoE links are brought up in the correct order, the system is free to boot however it wants. I don't see how this protocol needs to concern itself with "booting" versus "non-booting nodes.

Comment (atishp04) [reply InitializationProtocol/ManagementProtocol.adoc:37]: It depends on the scope of this protocol. I thought this protocol would specify the configuration of each node and complete booting sequence. We do need to specify all the possible states that nodes can be configured if we are talking about configuration.

If the scope only involves initialization, we probably don't need it.

File: InitializationProtocol/ManagementProtocol.adoc (hunk 1 above)
Comment (atishp04) [reply InitializationProtocol/ManagementProtocol.adoc:19]: :)

**Pull request closed without merging**